kytos/core/link.py

```diff
@@ -121,65 +123,55 @@ def id(self):  # pylint: disable=invalid-name
         """
         return self._id
 
-    @property
-    def available_tags(self):
+    def available_tags(self, tag_type: str = 'vlan') -> list[list[int]]:
         """Return the available tags for the link.
 
         Based on the endpoint tags.
         """
-        return [tag for tag in self.endpoint_a.available_tags if tag in
-                self.endpoint_b.available_tags]
-
-    def use_tag(self, tag):
-        """Remove a specific tag from available_tags if it is there.
-
-        Deprecated: use only the get_next_available_tag method.
-        """
-        if self.is_tag_available(tag):
-            self.endpoint_a.use_tag(tag)
-            self.endpoint_b.use_tag(tag)
-            return True
-        return False
+        return Interface.range_intersection(
+            self.endpoint_a.available_tags[tag_type],
+            self.endpoint_b.available_tags[tag_type],
+        )
 
-    def is_tag_available(self, tag):
+    def is_tag_available(self, tag: int, tag_type: str = 'vlan'):
         """Check if a tag is available."""
-        return (self.endpoint_a.is_tag_available(tag) and
-                self.endpoint_b.is_tag_available(tag))
-
-    def get_next_available_tag(self):
+        return (self.endpoint_a.is_tag_available(tag, tag_type) and
+                self.endpoint_b.is_tag_available(tag, tag_type))
+
+    def get_next_available_tag(
+        self,
+        controller,
+        tag_type: str = 'vlan'
+    ) -> int:
         """Return the next available tag if exists."""
         with self._get_available_vlans_lock:
-            # Copy the available tags because in case of error
-            # we will remove and add elements to the available_tags
-            available_tags_a = self.endpoint_a.available_tags.copy()
-            available_tags_b = self.endpoint_b.available_tags.copy()
-
-            for tag in available_tags_a:
-                # Tag does not exist in endpoint B. Try another tag.
-                if tag not in available_tags_b:
-                    continue
-
-                # Tag already in use. Try another tag.
-                if not self.endpoint_a.use_tag(tag):
-                    continue
-
-                # Tag already in use in B. Mark the tag as available again.
-                if not self.endpoint_b.use_tag(tag):
-                    self.endpoint_a.make_tag_available(tag)
-                    continue
-
-                # Tag used successfully by both endpoints. Returning.
-                return tag
+            with self.endpoint_a_lock:
+                with self.endpoint_b_lock:
+                    ava_tags_a = self.endpoint_a.available_tags[tag_type]
```

Comment:
Also, notice that in order to safely avoid the shallow copy without hitting a race condition, is to actually `acquire` the same underlying interfaces locks, so `self.endpoint_a._tag_lock` and `self.endpoint_b._tag_lock`, however `use_tags` as it is will try to reaquire so it will not work, so you'll need to parametrize a `lock=None` on `use_tags`, and if it's not none, then it should not acquire any lock with the assumption that the caller is already holding the lock, see what I mean?

Reply:
Added `use_lock` which is a boolean. If True then `use_tags` will use a Lock.

Reply:
@Alopalao, it's almost there, but instead of adding new locks references, the nested locks `self.endpoint_a._tag_lock` and `self.endpoint_b._tag_lock` are supposed to belong of the respective endpoint, reusing the existing ones just so they can lock accordingly. The lock parameters were as expected though. If you could double check this and then test it again ensuring that it'll lock as expected, that'd be great.

Reply:
Oh, right. Modified locks so they are from `Interface`